docs/design/crd-proposal-phase1.md

```diff
@@ -0,0 +1,220 @@
+
+# Introduction
+
+This document is a proposal for changes and enhancements of the
+CRDs (Custom Resource Defintions) of the operator. I'm calling the
+current state of the CRDs "Phase 0" and the next set of CRDs "Phase 1".
+
+The current, Phase 0, CRDs support two features. One can define a
+SmbService CR that specifies a PVC name and it will create services
+to export one share, named "share", based on the PVC name. One can
+define an SmbPvc CR that embeds the PVC spec. This will automatically
+create a PVC and a matching SmbService CR to export it.
+
+
+# Phase 1 Custom Resources
+
+The following CRD types are proposed for Phase 1:
+* SmbShare - A CR that encapsulates all the information needed to export a single share
+* SmbSecurityDomain - A CR that encapsulates the knowledge needed to create "local" users or bind to an Active Directory domain
+
+The operator will take the SmbShare and SmbSecurityDomain resources as inputs
+and create as many smbd, winbind, etc. backing services as needed. Users
+will have limited input into what backing services the operator will create.
+The operator may, or may not, combine one or more share into a single smbd
+instance.
+
+One or more SmbSecurityDomain resources can be defined in the cluster.  Each
+SmBShare CR can refer to one of those SmbSecurityDomain resources, or rely on
+the default settings of the operator.  The SmbSecurityDomain reference will
+define the security properties of the smbd instance that hosts the share. If
+two SmbShares are defined and each one refers to different SmbSecurityDomains
+they must not be combined under one smbd.
+
+The listings below are not meant to be entirely complete but they outline
+the genearal direction to make the operaterator a fully-fledged tool to
+create shares that can be consumed by a variety of clients for various use
+cases. Some sections are deliberately marked "TBD" (to be decided) as
+placeholders for more detail that will be needed even in Phase 1.
+
+## SmbSecurityDomain
+
+Spec Options:
+* `mode` - enumerated string "user" or "active-directory" - Indicates that this is an AD joined instance or based on a local mapping of users
+* `users` - mapping - A new subsection pertaining to "user" options - required when mode = "user"
+  * `secret` - string - The name of a secret containing user and group definitions that will be used to create the local users and groups for the service
```

Comment:
What exactly does the 'secret' string refer to? Is it some sort of embedding user and group definitions in a  cryptographical manner?

Reply:
In kubernets a secret is a bundle of configuration that may (or may not) be protected by encyption. By default it is simply obscured but you can install add ons that can protect the secret contents more thoroughly.